From e150d6e26092d8719d54c789882c63baf7348983 Mon Sep 17 00:00:00 2001
Message-Id: <e150d6e26092d8719d54c789882c63baf7348983.1298580377.git.jdenemar@redhat.com>
From: Eric Blake <eblake@redhat.com>
Date: Mon, 21 Feb 2011 12:30:41 -0700
Subject: [PATCH] network: plug unininitialized read found by valgrind

https://bugzilla.redhat.com/show_bug.cgi?id=679164

* src/util/network.c (virSocketAddrMask): Zero out port, so that
iptables can initialize just the netmask then call
virSocketFormatAddr without an uninitialized read in getnameinfo.
(cherry picked from commit 243b7814c778e555e63d105dac049291475221dc)
---
 src/util/network.c |    6 +++++-
 1 files changed, 5 insertions(+), 1 deletions(-)

diff --git a/src/util/network.c b/src/util/network.c
index a7e7423..33028aa 100644
--- a/src/util/network.c
+++ b/src/util/network.c
@@ -1,7 +1,7 @@
 /*
  * network.c: network helper APIs for libvirt
  *
- * Copyright (C) 2009-2010 Red Hat, Inc.
+ * Copyright (C) 2009-2011 Red Hat, Inc.
  *
  * See COPYING.LIB for the License of this software
  *
@@ -291,6 +291,7 @@ int virSocketAddrIsNetmask(virSocketAddrPtr netmask) {
  * virSocketAddrMask:
  * @addr: address that needs to be masked
  * @netmask: the netmask address
+ * @network: where to store the result, can be same as @addr
  *
  * Mask off the host bits of @addr according to @netmask, turning it
  * into a network address.
@@ -311,6 +312,7 @@ virSocketAddrMask(const virSocketAddrPtr addr,
         network->data.inet4.sin_addr.s_addr
             = (addr->data.inet4.sin_addr.s_addr
                & netmask->data.inet4.sin_addr.s_addr);
+        network->data.inet4.sin_port = 0;
         network->data.stor.ss_family = AF_INET;
         network->len = addr->len;
         return 0;
@@ -322,6 +324,7 @@ virSocketAddrMask(const virSocketAddrPtr addr,
                 = (addr->data.inet6.sin6_addr.s6_addr[ii]
                    & netmask->data.inet6.sin6_addr.s6_addr[ii]);
         }
+        network->data.inet6.sin6_port = 0;
         network->data.stor.ss_family = AF_INET6;
         network->len = addr->len;
         return 0;
@@ -334,6 +337,7 @@ virSocketAddrMask(const virSocketAddrPtr addr,
  * virSocketAddrMaskByPrefix:
  * @addr: address that needs to be masked
  * @prefix: prefix (# of 1 bits) of netmask to apply
+ * @network: where to store the result, can be same as @addr
  *
  * Mask off the host bits of @addr according to @prefix, turning it
  * into a network address.
-- 
1.7.4.1

